File: plugin-rust-agent/src/test/java/jetbrains/buildServer/rust/test/CargoLoggingListenerTest.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust.test;

import jetbrains.buildServer.agent.BuildProgressLogger;
import jetbrains.buildServer.rust.logging.CargoLoggerFactory;
import jetbrains.buildServer.rust.logging.CargoLoggingListener;
import jetbrains.buildServer.util.FileUtil;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.jetbrains.annotations.NotNull;
import org.jmock.Expectations;
import org.jmock.Mockery;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Dmitry.Tretyakov
 *         Date: 21.06.2016
 *         Time: 0:05
 */
public class CargoLoggingListenerTest {

    @Test(dataProvider = "getTestTransformations")
    public void testTransformation(final String filename) throws IOException {
        final File originalFile = new File("src/test/resources/cargo/original/" + filename);
        final List<String> original = new ArrayList<String>(FileUtil.readFile(originalFile));
        final File processedFile = new File("src/test/resources/cargo/processed/" + filename);
        final List<String> processed = new ArrayList<String>(FileUtil.readFile(processedFile));
        final Mockery m = new Mockery();
        final BuildProgressLogger logger = m.mock(BuildProgressLogger.class);
        final Matcher<String> matcher = getMatcher(processed);
        final int numberOfLines = processed.size();
        m.checking(new Expectations() {{
            exactly(numberOfLines).of(logger).message(with(matcher));
        }});

        final CargoLoggerFactory loggerFactory = new CargoLoggerFactory(logger);
        final CargoLoggingListener listener = new CargoLoggingListener(loggerFactory);

        for (String message : original) {
            listener.onStandardOutput(message);
        }

        listener.processFinished(0);

        m.assertIsSatisfied();
    }

    private BaseMatcher<String> getMatcher(@NotNull final List<String> processed) {
        return new BaseMatcher<String>() {
            @Override
            public boolean matches(Object actual) {
                return processed.size() > 0 && isEquals((String) actual, processed.remove(0));
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("List matcher");
            }
        };
    }

    private boolean isEquals(String actual, String expected) {
        actual = actual.replaceAll("\\s(duration='\\d+')", "");
        return actual.equals(expected);
    }

    @DataProvider
    public static Object[][] getTestTransformations() {
        return new Object[][]{
                {"cargoTests1.txt"}
        };
    }
}

File: plugin-rust-agent/src/test/kotlin/jetbrains/buildServer/rust/test/CargoLoggingListenerTest.kt
package jetbrains.buildServer.rust.test

import jetbrains.buildServer.agent.BuildProgressLogger
import jetbrains.buildServer.rust.logging.CargoLoggerFactory
import jetbrains.buildServer.rust.logging.CargoLoggingListener
import jetbrains.buildServer.util.FileUtil
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.jmock.Expectations
import org.jmock.Mockery
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import java.io.File
import java.io.IOException
import java.util.*

class CargoLoggingListenerTest {
    @Test(dataProvider = "testTransformations")
    @Throws(IOException::class)
    fun testTransformation(filename: String) {
        val originalFile = File("src/test/resources/cargo/original/$filename")
        val original = ArrayList(FileUtil.readFile(originalFile))
        val processedFile = File("src/test/resources/cargo/processed/$filename")
        val processed = ArrayList(FileUtil.readFile(processedFile))
        val m = Mockery()
        val logger = m.mock<BuildProgressLogger>(BuildProgressLogger::class.java)
        val matcher = getMatcher(processed)
        val numberOfLines = processed.size
        m.checking(object : Expectations() {
            init {
                exactly(numberOfLines).of(logger).message(with(matcher))
            }
        })

        val loggerFactory = CargoLoggerFactory(logger)
        val listener = CargoLoggingListener(loggerFactory)

        for (message in original) {
            listener.onStandardOutput(message)
        }

        listener.processFinished(0)

        m.assertIsSatisfied()
    }

    private fun getMatcher(processed: MutableList<String>): BaseMatcher<String> {
        return object : BaseMatcher<String>() {
            override fun matches(actual: Any): Boolean =
                processed.isNotEmpty() && isEquals(actual as String, processed.removeAt(0))

            override fun describeTo(description: Description) {
                description.appendText("List matcher")
            }
        }
    }

    private fun isEquals(actual: String, expected: String): Boolean {
        val cleaned = actual.replace("\\s(duration='\\d+')".toRegex(), "")
        return cleaned == expected
    }

    @DataProvider
    fun testTransformations(): Array<Array<Any>> = arrayOf(
        arrayOf("cargoTests.txt"),
        arrayOf("cargoCompileVerbose.txt"),
        arrayOf("cargoTestFailures.txt"),
        arrayOf("cargoBuild.txt"),
        arrayOf("cargoWarnings.txt"),
        arrayOf("cargoDocTests1.txt"),
        arrayOf("cargoDocTests2.txt"),
        arrayOf("cargoDocTests3.txt"),
        arrayOf("cargoDocTests4.txt"),
        arrayOf("cargoFeatureError.txt")
    )
}
